### ISSUE TITLE
Microsoft.Azure.Devices nuget package v 1.1.2 has bad reference to Microsoft.Azure.Devices.Shared assembly

### ISSUE THREAD

**Issue by obsoleted** [Azure/azure-iot-sdks#1009]
The Microsoft.Azure.Devices assembly in the 1.1.2 nuget package contains a bad reference to Microsoft.Azure.Devices.Shared. The reference appears to be non-strongkey signed version. When using the sdk the below exception is seen because only a strong name version is found which doesn't match.

```
System.IO.FileLoadException was unhandled
Message: An unhandled exception of type 'System.IO.FileLoadException' occurred in mscorlib.dll
Additional information: Could not load file or assembly 'Microsoft.Azure.Devices.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. A strongly-named assembly is required. (Exception from HRESULT: 0x80131044)
```

Repro code for above:

```C#
            var sc = Microsoft.Azure.Devices.RegistryManager.CreateFromConnectionString("...");

            var devices = sc.GetDevicesAsync(5).Result;

            foreach(var device in devices)
            {
                Console.WriteLine($"Device: {device.Id}");
            }
```

**Comment by olivierbloch** [comment]
Hi @obsoleted 
We are aware of the issue and are working on a resolution. stay tuned

**Comment by BertKleewein** [comment]
@obsoleted - you should be able to manually add a reference to Microsoft.Azure.Devices.Shared to unblock yourself until we can get an official fix out.

**Comment by obsoleted** [comment]
@BertKleewein I'm not exactly sure what you mean.  The app already has references to Shared in nuget and  in the project (a seemingly correct reference with a non null public key token value). And i can successfully touch and use types in that assembly directly.

In code below the app successfully constructs a Twin object and gets past the first WriteLine. But fails when getting to the function because it loads Microsoft.Azure.Devices which itself wants an unsigned Microsoft.Azure.Devices.Shared which it can't find (Same exception as above comment).

Anyway, we were able to get by using an older version of the package for simple use case.

Refactoring sample:

```C#
        static void Main(string[] args)
        {

            var twin = new Microsoft.Azure.Devices.Shared.Twin();
            Console.WriteLine($"Twin: {twin}"); // Execution gets here fine
            
            TouchMAD();
        }

        static void TouchMAD()
        {
            Console.WriteLine($"Try to enumerate devices...");
            var sc = Microsoft.Azure.Devices.RegistryManager.CreateFromConnectionString("...");

            var devices = sc.GetDevicesAsync(5).Result;

            foreach (var device in devices)
            {
                Console.WriteLine($"Device: {device.Id}");
            }
        }
```

App.csproj excerpt:
```
    <Reference Include="Microsoft.Azure.Devices.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
      <HintPath>..\packages\Microsoft.Azure.Devices.Shared.1.0.2\lib\net45\Microsoft.Azure.Devices.Shared.dll</HintPath>
      <Private>True</Private>
    </Reference>
```

Package.json excerpt:
```
  <package id="Microsoft.Azure.Devices" version="1.1.2" targetFramework="net452" />
  <package id="Microsoft.Azure.Devices.Client" version="1.1.1" targetFramework="net452" />
  <package id="Microsoft.Azure.Devices.Shared" version="1.0.2" targetFramework="net452" />
```

**Comment by BertKleewein** [comment]
@obsoleted  - are you build a UAP project?  I am currently investigating a signing issue when using this library with UAP and suspect that it is the same issue.

**Comment by BertKleewein** [comment]
@obsoleted - Thanks for the extra information.  I was looking at a similar problem in a different assembly and I assumed that they were related.  I suspect that I am probably wrong.  I am currently investigating this and will let you know what I find.

**Comment by obsoleted** [comment]
I was just making a normal console app but the UWP issue might be the same. (Though I'm not super familiar with the dependency loading in UWP/WinRT land w.r.t strong name signing).
Looking at the Microsoft.Azure.Devices assembly in ILSpy yields the following references. I've _italicized_ what is the problematic reference and **bolded** the badd PublicKeyToken value setting.

References for // Microsoft.Azure.Devices, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35

> // mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
> // Microsoft.Azure.Amqp, Version=1.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
> // System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
> // System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed
> *// **Microsoft.Azure.Devices.Shared**, Version=1.0.0.0, Culture=neutral, **PublicKeyToken=null***
> // System.Runtime.Serialization, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
> // System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
> // System.Transactions, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
> // Microsoft.Owin, Version=3.0.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
> // System.Configuration, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Web, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Net.Http.Formatting, Version=5.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
> // System.ServiceModel, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
> // advapi32
> // kernel32.dll
> // ws2_32.dll
> // advapi32.dll

References for // Microsoft.Azure.Devices.Uwp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35

> // System.Runtime, Version=4.0.20.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Runtime.InteropServices, Version=4.0.20.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // Microsoft.Azure.Amqp.Uwp, Version=1.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
> // System.Threading.Tasks, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Diagnostics.Debug, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.IO, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Collections, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Net.Primitives, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // Windows.Foundation.UniversalApiContract [WinRT]
> // System.Globalization, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Diagnostics.Tools, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Resources.ResourceManager, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed
> _// Microsoft.Azure.Devices.Shared.WinRT, Version=1.0.0.0, Culture=neutral, **PublicKeyToken=null**_
> // System.Runtime.Serialization.Primitives, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Net.WebHeaderCollection, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Text.RegularExpressions, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Linq.Expressions, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Reflection, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Collections.Concurrent, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Threading.Timer, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Runtime.Extensions, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.ObjectModel, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Threading, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // System.Text.Encoding, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> // PCLCrypto, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d4421c8a4786956c
> // System.Reflection.TypeExtensions, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
> 
> 


And again the version of the Microsoft.Azure.Devices.Shared assembly that nuget brings down **does** have proper signing i think:
// **Microsoft.Azure.Devices.Shared**, Version=1.0.0.0, Culture=neutral, **PublicKeyToken=31bf3856ad364e35**

**Comment by BertKleewein** [comment]
Thanks.  I think I know where the problem lies.  There are some scripts in our build environment that are causing this token to get lost.

I also learned a few more tools for this:
1. sn -T foo.dll will give you the strong name token for a particular DLL.  
2. you can use ildsasm to look at the assembly manifest to see what sn tokens that assembly is referencing.  

For example, this is what I expect to see on Microsoft.Azure.Device.dll;
```
.assembly extern Microsoft.Azure.Devices.Shared
{
  .publickeytoken = (31 BF 38 56 AD 36 4E 35 )                         // 1.8V.6N5
  .ver 1:0:0:0
}
```
but this is what I do see:
```
.assembly extern Microsoft.Azure.Devices.Shared
{
  .ver 1:0:0:0
}
```

I will start working on a fix.

(You probably knew this already, but I learned something new here, so, thank you).

**Comment by BertKleewein** [comment]
This is fixed in Microsoft.Azure.Devices v 1.1.3.  I'm going to close this issue for now, please re-activate (or open a new issue) if you still have problems.
